### PR TITLE
enhance(chips): Many updates to Chips

### LINKED ISSUE #2123: Chips input inconsistent style
Take a look at this screenshot, the input is very different from something like autocomplete.

screen shot 2015-04-01 at 6 05 21 pm

### LINKED ISSUE #2145: Chips accessibility improvements
Here are some things we can do to improve chips:

 * Put `tabindex="-1"` on `md-chips`, which is getting `tabindex="0"` from ngAria and ngModel.
 * The `role=presentation` on `md-chips-wrap > div` is unnecessary.
 * Make the chip one tab stop and wire up the DEL key to delete.
 * Put ng-click on the X with `tabIndex="-1"` to remove it from the tab order but allow clicks to remove.
 * The text in it can say something like "Apple delete" with the unique part coming first.
 * Removing a chip: Focus should be sent back to the parent `md-chips` element, so that focus is not dropped.
 * Input fields need `aria-label` values (or `<label for="">`, either on- or off-screen).

Question:
 * Should readonly chips have `tabindex`? 

Kudos for the existing keyboard support, this is shaping up to be an awesome component!

### PR DESCRIPTION
**Pull request by typotter**
- Brought style (closer) to spec
- support user pass mdautocomplete and input elements
- tests for mdchipappend
- update demo formatting, structure
- Aria Fixes #2145
- Style fixes #2123
- insert default md-remove when none provided
- make md-chip-remove a sibling to the chip content so that there is no nested ng-click
- refactor DSL API to use md-chip-template instead of md-chip
- support static chips
- delete key on selected chip deletes it

### CODE REVIEW

File: src/components/chips/chips.scss [hunk 1]
```diff
@@ -17,29 +17,37 @@ $chip-wrap-padding: 0.8rem !default;
     border-radius: $chip-height / 2;
     display: inline-block;
     height: $chip-height;
-    margin: $chip-margin;
     line-height: $chip-height;
+    margin: $chip-margin;
     padding: $chip-padding;
 
+    > div {
```
Comment (jelbourn) [inline src/components/chips/chips.scss]: Instead of a direct child, can you just add another class?

File: src/components/chips/chips.scss [hunk 2]
```diff
@@ -48,13 +56,46 @@ $chip-wrap-padding: 0.8rem !default;
...
+
     &:not(:first-child) {
       margin: $chip-margin;
     }
-  }
 
-  .md-chip-input {
-    background-color:transparent;
-    border-width: 0;
+    input {
```
Comment (jelbourn) [inline src/components/chips/chips.scss]: Might this to too broad?

File: src/components/chips/chips.spec.js [hunk 3]
```diff
@@ -82,16 +82,155 @@ describe('<md-chips>', function() {
       expect(chips[0].innerHTML).toContain('Apple');
       expect(chips[1].innerHTML).toContain('Orange');
     });
+
+    it('should call the append method when adding a chip', function() {
+      var element = buildChips('<md-chips ng-model="items" md-chip-append="appendChip($chip)"></md-chips>');
```
Comment (jelbourn) [inline src/components/chips/chips.spec.js]: Can you keep lines under 100 columns?

File: src/components/chips/js/chipDirective.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+   *
+   *
+   * @usage
+   * <hljs lang="html">
+   *   <md-chip>{{$chip}}</md-chip>
+   * </hljs>
+   *
+   */
+
+  var DELETE_HINT_TEMPLATE = '\
```
Comment (jelbourn) [inline src/components/chips/js/chipDirective.js]: Add a comment describing what this template is used for.

File: src/components/chips/js/chipDirective.js [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+      <span class="visually-hidden">Press delete to remove this chip</span>';
```
Comment (jelbourn) [inline src/components/chips/js/chipDirective.js]: This message can't be defined here; users need to be able to specify it so that it can be localized.

File: src/components/chips/js/chipRemoveDirective.js [hunk 6]
```diff
@@ -37,7 +37,14 @@
       restrict: 'E',
       template: REMOVE_CHIP_TEMPLATE,
       require: ['^mdChips'],
-      scope: false
+      scope: false,
+      link: postLink
     };
+
+    function postLink(scope, element, attr, ctrl) {
+      $timeout(function() {
```
Comment (jelbourn) [inline src/components/chips/js/chipRemoveDirective.js]: Add comment explaining why you use the `$timeout`

File: src/components/chips/js/chipsController.js [hunk 7]
```diff
@@ -64,27 +70,57 @@
...
-        if ( this.chipBuffer ) {
+        var chipBuffer = this.getChipBuffer();
+        if (chipBuffer) {
           event.preventDefault();
-          this.appendChipBuffer();
+          this.appendChip(chipBuffer);
+          this.resetChipBuffer();
         }
         break;
       case this.$mdConstant.KEY_CODE.BACKSPACE: // backspace
```
Comment (jelbourn) [inline src/components/chips/js/chipsController.js]: The `// backspace` comment is not needed

File: src/components/chips/js/chipsController.js [hunk 8]
```diff
@@ -64,27 +70,57 @@
...
 
 
   /**
+   * Handles the keydown event on an `<md-chip>` element.
+   * @param index
+   * @param event
+   */
+  MdChipsCtrl.prototype.chipKeydown = function(index, event) {
+    switch (event.keyCode) {
+      case this.$mdConstant.KEY_CODE.DELETE: // backspace
```
Comment (jelbourn) [inline src/components/chips/js/chipsController.js]: Delete and backspace should both probably do the same thing here.

File: src/components/chips/js/chipsController.js [hunk 9]
```diff
@@ -93,19 +129,38 @@
...
+    var len = this.items.length - 1;
+    if (this.selectChip != index || len == 0) {
+      return -1;
+    } else if (index == len) {
+      return index - 1;
+    } else {
+      return index;
+    }
+  };
+
```
Comment (jelbourn) [inline src/components/chips/js/chipsController.js]: You don't need `else` in this function:
```js
var len = this.items.length - 1;
if (this.selectChip != index || len == 0) {
  return -1;
} 

if (index == len) {
  return index - 1;
} 

return index;
```

File: src/components/chips/js/chipsController.js [hunk 10]
```diff
@@ -93,19 +129,38 @@
...
+  /**
    * Append the contents of the buffer to the chip list. This method will first
    * call out to the md-chip-append method, if provided
+   * @param newChip
    */
-  MdChipsCtrl.prototype.appendChipBuffer = function() {
-    var newChip = this.getChipBuffer();
-    if (this.useMdChipAppend && this.mdChipAppend) {
+  MdChipsCtrl.prototype.appendChip = function(newChip) {
+    if (this.useMdChipAppend && !!this.mdChipAppend) {
```
Comment (jelbourn) [inline src/components/chips/js/chipsController.js]: You never need `!!` in conditionals like `if`, `while`, etc. You only would use `!!` if you have a function that returns or is expecting a `boolean` and nothing else.

File: src/components/chips/js/chipsController.js [hunk 11]
```diff
@@ -193,6 +260,41 @@
    */
   MdChipsCtrl.prototype.configureMdAutocomplete = function(mdAutocompleteCtrl) {
     this.mdAutocompleteCtrl = mdAutocompleteCtrl;
-    // TODO(typotter): create and register a selectedItem watcher with mdAutocompleteCtrl.
+    this.mdAutocompleteCtrl.registerSelectedItemWatcher(angular.bind(this, this.mdAutocompleteSelectedItemWatcher));
+  };
+
+
+  MdChipsCtrl.prototype.mdAutocompleteSelectedItemWatcher = function(newItem, oldItem) {
+    if (!!newItem && newItem !== oldItem) {
```
Comment (jelbourn) [inline src/components/chips/js/chipsController.js]: You never need !! in conditionals like if, while, etc. You only would use !! if you have a function that returns or is expecting a boolean and nothing else.

File: src/components/chips/js/chipsController.js [hunk 12]
```diff
@@ -193,6 +260,41 @@
    */
   MdChipsCtrl.prototype.configureMdAutocomplete = function(mdAutocompleteCtrl) {
     this.mdAutocompleteCtrl = mdAutocompleteCtrl;
-    // TODO(typotter): create and register a selectedItem watcher with mdAutocompleteCtrl.
+    this.mdAutocompleteCtrl.registerSelectedItemWatcher(angular.bind(this, this.mdAutocompleteSelectedItemWatcher));
```
Comment (jelbourn) [inline src/components/chips/js/chipsController.js]: You can just use `this.mdAutocompleteSelectedItemWatcher.bind(this)`

File: src/components/chips/js/chipsController.js [hunk 13]
```diff
@@ -128,9 +183,13 @@
...
+      // Should not be called from here.
+    } else if (this.userInputElement) {
+      if (this.userInputNgModelCtrl) {
+        return this.userInputNgModelCtrl.$viewValue;
+      } else {
+        return this.userInputElement[0].value;
+      }
     } else {
       return this.chipBuffer;
     }
```
Comment (jelbourn) [inline src/components/chips/js/chipsController.js]: Can eliminate some `else` here:

```js
if (this.mdAutocompleteCtrl) {
  throw Error('...');
}

if (this.userInputElement) {
  return this.userInputNgModelCtrl ? 
      this.userInputNgModelCtrl.$viewValue : 
      this.userInputElement[0].value;
}

return this.chipsBuffer;
```

File: src/components/chips/js/chipsDirective.js [hunk 14]
```diff
@@ -136,49 +142,99 @@
       }
     };
     function compile(element, attr) {
-      var userTemplate = attr['$mdUserTemplate'];
-      var chipEl = userTemplate.find('md-chip');
-      var chipHtml;
-      if (chipEl.length === 0) {
+      function getWorkerChip(el) {
```
Comment (jelbourn) [inline src/components/chips/js/chipsDirective.js]: What's a worker chip?

File: src/components/chips/js/chipsDirective.js [hunk 15]
```diff
@@ -136,49 +142,99 @@
...
+        return angular.element(el[0].querySelector('.md-chip-worker'));
+      }
+
+      var chipHtml,
+          userTemplate = attr['$mdUserTemplate'],
```
Comment (jelbourn) [inline src/components/chips/js/chipsDirective.js]: If you don't use it any more after this, do
`attr['$mdUserTemplate'] = null;`

File: src/components/chips/js/chipsDirective.js [hunk 16]
```diff
@@ -136,49 +142,99 @@
...
+          chipTemplate.remove('md-chip-remove');
         }
-        // Take only the chip's inner HTML as the encasing repeater is an md-chip element.
-        chipHtml = chipEl[0].innerHTML;
+        chipHtml = chipTemplate[0].innerHTML;
       }
-      var listNode = angular.element(element[0].querySelector('.md-chip'));
-      listNode.append(chipHtml);
+
+      var chipContentNode = angular.element(element[0].querySelector('md-chip div'));
```
Comment (jelbourn) [inline src/components/chips/js/chipsDirective.js]: The selector `md-chip div` is really brittle.

File: src/components/chips/js/chipsDirective.js [hunk 17]
```diff
@@ -136,49 +142,99 @@
...
+        hasInputElement = userInput.length > 0;
 
-        // Default element.
-        inputEl = angular.element(CHIP_INPUT_TEMPLATE);
-        var workerChip = angular.element(element[0].querySelector('.md-chip-worker'));
-        workerChip.append(inputEl);
+        if (hasInputElement) {
+          transcludeInputElement = userInput[0];
+        } else {
+          // Only add the input element if there is ngModel.
```
Comment (jelbourn) [inline src/components/chips/js/chipsDirective.js]: I don't quite follow what's going on here.

File: src/components/chips/js/chipsController.js (hunk 13 above)
Comment (ThomasBurleson) [inline src/components/chips/js/chipsController.js]: This version is much more clear:


```js
if (this.mdAutocompleteCtrl) {
  throw Error('...');
}

return !this.userInputElement ? this.chipsBuffer :
		this.userInputNgModelCtrl ? this.userInputNgModelCtrl.$viewValue : 
      	this.userInputElement[0].value;
```

Almost always best to use single-return at bottom of function.

> sometimes may use return at top for **rapid exits**; aka `returning early`

Comment (jelbourn) [inline src/components/chips/js/chipsController.js]: Personally I find the nested ternary operators much less readable.

**Comment by ThomasBurleson** [comment]
@typotter  - is this ready and adjustments based on feedback ?

File: src/components/chips/chips.scss (hunk 1 above)
Comment (typotter) [inline src/components/chips/chips.scss]: done

File: src/components/chips/chips.scss (hunk 2 above)
Comment (typotter) [inline src/components/chips/chips.scss]: no. Want any inputs as children here to conform to this style

File: src/components/chips/chips.spec.js (hunk 3 above)
Comment (typotter) [inline src/components/chips/chips.spec.js]: done

File: src/components/chips/js/chipDirective.js (hunk 4 above)
Comment (typotter) [inline src/components/chips/js/chipDirective.js]: done

File: src/components/chips/js/chipDirective.js (hunk 5 above)
Comment (typotter) [inline src/components/chips/js/chipDirective.js]: Still need to do this for this message and the delete button hint. Would it be acceptable to define a default value for these messages and allow them to be overridden via attributes on md-chips?

File: src/components/chips/js/chipRemoveDirective.js (hunk 6 above)
Comment (typotter) [inline src/components/chips/js/chipRemoveDirective.js]: done

File: src/components/chips/js/chipsController.js (hunk 7 above)
Comment (typotter) [inline src/components/chips/js/chipsController.js]: done

File: src/components/chips/js/chipsController.js (hunk 8 above)
Comment (typotter) [inline src/components/chips/js/chipsController.js]: done

File: src/components/chips/js/chipsController.js (hunk 9 above)
Comment (typotter) [inline src/components/chips/js/chipsController.js]: done

File: src/components/chips/js/chipsController.js (hunk 11 above)
Comment (typotter) [inline src/components/chips/js/chipsController.js]: done

File: src/components/chips/js/chipsController.js (hunk 12 above)
Comment (typotter) [inline src/components/chips/js/chipsController.js]: done

File: src/components/chips/js/chipsController.js (hunk 13 above)
Comment (typotter) [inline src/components/chips/js/chipsController.js]: done

File: src/components/chips/js/chipsDirective.js (hunk 14 above)
Comment (typotter) [inline src/components/chips/js/chipsDirective.js]: vestige of a layout when the input control was inside a chip. renamed to input container

File: src/components/chips/js/chipsDirective.js (hunk 15 above)
Comment (typotter) [inline src/components/chips/js/chipsDirective.js]: done

File: src/components/chips/js/chipsDirective.js (hunk 16 above)
Comment (typotter) [inline src/components/chips/js/chipsDirective.js]: replaced w/class selector

File: src/components/chips/js/chipsDirective.js (hunk 17 above)
Comment (typotter) [inline src/components/chips/js/chipsDirective.js]: rearranged the code a bit and added comments.

**Comment by typotter** [comment]
@ThomasBurleson yes. Still have a small TODO wrt hint text, but will push that change in a short while.

**Comment by typotter** [comment]
@ThomasBurleson just rebased and push a change w/the hint text attributes. Will squash and rebase again once you approve the changes

File: src/components/chips/chips.scss (hunk 2 above)
Comment (jelbourn) [inline src/components/chips/chips.scss]: Even checkboxes? Radio buttons?

File: src/components/chips/chips.spec.js [hunk 18]
```diff
@@ -1,24 +1,28 @@
...
       var ctrl = element.controller('mdChips');
 
       var input = element.find('input');
       expect(input.length).toBe(1);
-      expect(input).toHaveClass('md-chip-input');
     });
 
     it('should render a list of chips', function() {
-      var element = buildChips( '<md-chips ng-model="items"></md-chips>' );
+      var element = buildChips( BASIC_CHIP_TEMPLATE );
```
Comment (jelbourn) [inline src/components/chips/chips.spec.js]: Inconsistent spacing around function arguments. I'd say to just trim them.

File: src/components/chips/js/chipDirective.js [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+   * <hljs lang="html">
+   *   <md-chip>{{$chip}}</md-chip>
+   * </hljs>
+   *
+   */
+
+  // This hint text is hidden within a chip but used by screen readers to
+  // inform the user how they can interact with a chip.
+  var DELETE_HINT_TEMPLATE = '\
+      <span ng-if="!$mdChipsCtrl.readonly" class="visually-hidden">\
```
Comment (jelbourn) [inline src/components/chips/js/chipDirective.js]: Shouldn't this `visually-hidden` class be prefixed with `md-` ?

File: src/components/chips/js/chipDirective.js [hunk 20]
```diff
@@ -0,0 +1,54 @@
...
+        {{$mdChipsCtrl.deleteHint || $mdChipsCtrl.defaultDeleteHint}}\
```
Comment (jelbourn) [inline src/components/chips/js/chipDirective.js]: Instead of putting an `||` in the binding expression, just default `deleteHint` to the default message.

File: src/components/chips/js/chipsController.js [hunk 21]
```diff
@@ -128,21 +215,25 @@
...
+    return !this.userInputElement ? this.chipBuffer :
+        this.userInputNgModelCtrl ? this.userInputNgModelCtrl.$viewValue :
+            this.userInputElement[0].value;
   };
 
 
   /**
-   * Resets the input buffer.
+   * Resets the input buffer be it the internal input or user provided input element.
```
Comment (jelbourn) [inline src/components/chips/js/chipsController.js]: `Resets the input buffer for either the internal input or the user provided input element.`

File: src/components/chips/js/chipsController.js [hunk 22]
```diff
@@ -193,6 +324,36 @@
...
+    if (newItem && newItem !== oldItem) {
+      this.appendChip(newItem);
+      this.mdAutocompleteCtrl.clear();
+    }
+  };
+
+
+  /**
+   * Configure event bindings on a user-provided input element.
+   * @param inputElement {!angular.Element}
```
Comment (jelbourn) [inline src/components/chips/js/chipsController.js]: JsDoc syntax is `@param {type} varName`

File: src/components/chips/js/chipsDirective.js [hunk 23]
```diff
@@ -120,65 +148,161 @@
...
+     * Static Chips (i.e. `md-chip` elements passed from the caller) are gathered and set aside for
+     * transclusion in the `postLink` function.
+     *
+     *
+     * @param element
+     * @param attr
+     * @returns {Function}
+     */
     function compile(element, attr) {
+      function getInputContainer(el) {
```
Comment (jelbourn) [inline src/components/chips/js/chipsDirective.js]: This function should be defined outside the `compile` function.

File: src/components/chips/js/chipsController.js [hunk 24]
```diff
@@ -64,27 +82,81 @@
...
+   * Get the input's placeholder - uses `placeholder` when list is empty and `secondary-placeholder`
+   * when the list is non-empty. If `secondary-placeholder` is not provided, `placeholder` is used
+   * always.
+   */
+  MdChipsCtrl.prototype.getPlaceholder = function() {
+    return this.items.length == 0 ?
+        this.placeholder :
+        this.secondaryPlaceholder ?
+            this.secondaryPlaceholder :
+            this.placeholder;
```
Comment (jelbourn) [inline src/components/chips/js/chipsController.js]: I find nested ternary expressions like this incredibly hard to read...

Comment (ThomasBurleson) [inline src/components/chips/js/chipsController.js]: In fact that logic is confusing in general. 
Here is an improved version:

```js
return (this.items.length && this.secondaryPlaceholder) ? this.secondaryPlaceholder : this.placeholder;
```

File: src/components/chips/js/chipDirective.js (hunk 19 above)
Comment (typotter) [inline src/components/chips/js/chipDirective.js]: yes, probably. prefer to do refactor in its own pr.

File: src/components/chips/chips.scss (hunk 2 above)
Comment (typotter) [inline src/components/chips/chips.scss]: hmm. I will limit the style to text and number for now. We can add other input types as needed
